### DISCUSSION TITLE
Function idea: bru_ds() for easy fitting of key functions and adjustments

### DISCUSSION THREAD

**Discussion by andyhouldcroft** [inlabru-org/inlabru#255] [Ideas]
Hi all,

I think this has been suggested before, but i think it would be really valuable for inlabru to have a user-friendly function (e.g., bru_ds) that simplifies the defining of the three key distance sampling functions (uniform, half-normal, hazard-rate) and their associated adjustments (cosine, simple polynomial, and hermite polynomial).

I am fairly inexperienced in writing R functions but i have included an incomplete starting point below if it is of any use! There is a comment included to indicate where adjustment code could be added, i just wasn't exactly sure on the implementation of the adjustment terms.

All the best,
Andrew

```
bru_ds <- function(distance, sigma, gamma = NULL, width = NULL, key = "hn", adjust = NULL, order = NULL, log = TRUE) {
  
  if (!key %in% c("unif", "hn", "hr")) {
    stop("Invalid 'key' value. Choose from 'unif', 'hn', or 'hr'.")
  }
  
  if (!is.null(adjust) && !adjust %in% c("herm", "cos", "poly")) {
    stop("Invalid 'adjust' value. Choose from 'herm', 'cos', or 'poly'.")
  }
  
  if (missing(sigma)) {
    stop("The parameter 'sigma' is required.")
  }
  
  if (key == "hr" && missing(gamma)) {
    stop("The parameter 'gamma' is required when 'key' is set to 'hr'.")
  }
  
  if (key == "unif" && missing(width)) {
    stop("The parameter 'width' is required when 'key' is set to 'unif'.")
  }
  
  if (!is.null(adjust) && is.null(order)) {
    stop("The parameter 'order' is required when 'adjust' is specified.")
  }
  
  valid_adjustments <- list(
    unif = c("cos", "poly"),
    hn = c("herm", "cos"),
    hr = c("cos", "poly")
  )
  
  if (!is.null(adjust) && !adjust %in% valid_adjustments[[key]]) {
    stop(sprintf("The adjustment '%s' is not valid for key '%s'. Valid adjustments for '%s' are: %s.", 
                 adjust, key, key, paste(valid_adjustments[[key]], collapse = ", ")))
  }
  
  key_list <- list(
    unif = function(x) 1 / width,
    hn = function(x) exp(-0.5 * (x / sigma)^2),
    hr = function(x) 1 - exp(-(x / sigma)^-gamma)
  )
  
  det_prob <- key_list[[key]](distance)
  
  if (!is.null(adjust)) det_prob <- # add here?
  
  if (log) det_prob <- log(det_prob)
  return(det_prob)
  }
```

**Comment by finnlindgren** [comment]
Yeah, something like this has been thought of for quite a while; see the enhancement issue raised in #4. Also see the discussion in #195 about some of the semi-parametric adjustment methods, showing how to implement a cosine sum adjustment (and noting that the way it's written in the literature is either not what's actually done, or they deal with the resulting negative probabilities in some way; the discussion example shows how to apply the adjustment additively on the log-scale, which is well-defined).
To implement these methods in a well-defined way, I'd need some good references to what the models actually are (and some might then not be implementable, or implementable as different models but similar in spirit).

**Reply by finnlindgren** [reply]
Speaking of well-definedness, what is "unif" supposed to represent, in terms of detection probability? It looks more like a density model (and 1/W isn't restricted to [0,1]), and as such not really a part of the detection model.
